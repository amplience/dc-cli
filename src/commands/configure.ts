import { Arguments } from 'yargs';
import { CommandOptions } from '../interfaces/command-options.interface';
import fs from 'fs';
import { join, dirname } from 'path';
import { isEqual } from 'lodash';

export const command = 'configure';

export const desc = 'Saves the configuration options to a file';

export const CONFIG_FILENAME = (platform: string = process.platform): string =>
  join(process.env[platform == 'win32' ? 'USERPROFILE' : 'HOME'] || __dirname, '.amplience', 'dc-cli-config.json');

export type ConfigurationParameters = {
  clientId: string;
  clientSecret: string;
  hubId: string;
};

type ConfigArgument = {
  config: string;
};

export const configureCommandOptions: CommandOptions = {
  clientId: { type: 'string', demandOption: true },
  clientSecret: { type: 'string', demandOption: true },
  hubId: { type: 'string', demandOption: true },
  config: { type: 'string', default: CONFIG_FILENAME() }
};

const writeConfigFile = (configFile: string, parameters: ConfigurationParameters): void => {
  const dir = dirname(configFile);
  if (!fs.existsSync(dir)) {
    try {
      fs.mkdirSync(dir, { recursive: true });
    } catch (err) {
      throw new Error(`Unable to create dir "${dir}". Reason: ${err}`);
    }
  }
  try {
    fs.writeFileSync(configFile, JSON.stringify(parameters));
  } catch (err) {
    throw new Error(`Unable to write config file "${configFile}". Reason: ${err}`);
  }
};

export const readConfigFile = (configFile: string): object =>
  fs.existsSync(configFile) ? JSON.parse(fs.readFileSync(configFile, 'utf-8')) : {};

export const handler = (argv: Arguments<ConfigurationParameters & ConfigArgument>): void => {
  const { clientId, clientSecret, hubId } = argv;
  const storedConfig = readConfigFile(argv.config);

  if (isEqual(storedConfig, { clientId, clientSecret, hubId })) {
    console.log('Config file up-to-date.  Please use `--help` for command usage.');
    return;
  }
  writeConfigFile(argv.config, { clientId, clientSecret, hubId });
  console.log('Config file updated.');
};
