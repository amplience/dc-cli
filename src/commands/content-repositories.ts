import { Argv } from 'yargs';
import YargsCommandBuilderOptions from '../common/yargs/yargs-command-builder-options';

export const command = 'content-repositories';

export const desc = 'Content Repositories';

export const builder = (yargs: Argv): Argv =>
  yargs
    .commandDir('content-repositories', YargsCommandBuilderOptions)
    .demandCommand()
    .help();

export const handler = (): void => {};
