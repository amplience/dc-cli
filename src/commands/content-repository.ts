import { Argv } from 'yargs';
import YargsCommandBuilderOptions from '../common/yargs/yargs-command-builder-options';

export const command = 'content-repository';

export const desc = 'Content Repository';

export const builder = (yargs: Argv): Argv =>
  yargs
    .commandDir('content-repository', YargsCommandBuilderOptions)
    .demandCommand()
    .help();

export const handler = (): void => {
  /* do nothing */
};
