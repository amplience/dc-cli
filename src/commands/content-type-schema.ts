import { Argv } from 'yargs';
import YargsCommandBuilderOptions from '../common/yargs/yargs-command-builder-options';

export const command = 'content-type-schema';

export const desc = 'Content Type Schema';

export const builder = (yargs: Argv): Argv =>
  yargs
    .commandDir('content-type-schema', YargsCommandBuilderOptions)
    .demandCommand()
    .help();

export const handler = (): void => {
  /* do nothing */
};
