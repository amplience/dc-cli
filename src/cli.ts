import Yargs from 'yargs/yargs';
import YargsCommandBuilderOptions from './common/yargs/yargs-command-builder-options';
import { configureCommandOptions, readConfigFile } from './commands/configure';
import { Arguments, Argv } from 'yargs';
import errorHandler from './error-handler';

const configureYargs = (yargInstance: Argv): Promise<Arguments> => {
  return new Promise(
    async (resolve): Promise<void> => {
      let failInvoked = false;
      const isYError = (err?: Error | string): boolean => err instanceof Error && err.name === 'YError';
      const failFn = (msg: string, err?: Error | string): void => {
        // fail should only be invoked once
        if (failInvoked) {
          return;
        }
        failInvoked = true;
        if ((msg && !err) || isYError(err)) {
          yargInstance.showHelp('error');
        }
        errorHandler(err || msg);
      };
      const argv = await yargInstance
        .scriptName('dc-cli')
        .options(configureCommandOptions)
        .config('config', readConfigFile)
        .commandDir('./commands', YargsCommandBuilderOptions)
        .strict()
        .demandCommand(1, 'Please specify at least one command')
        .exitProcess(false)
        .showHelpOnFail(false)
        .fail(failFn).argv;
      resolve(argv);
    }
  );
};

export default async (yargInstance = Yargs(process.argv.slice(2))): Promise<Arguments | void> => {
  return await configureYargs(yargInstance);
};
