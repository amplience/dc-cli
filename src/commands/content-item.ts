import { Argv } from 'yargs';
import { readConfig } from '../cli';
import YargsCommandBuilderOptions from '../common/yargs/yargs-command-builder-options';
import { configureCommandOptions } from './configure';

export const command = 'content-item';

export const desc = 'Content Item';

export const builder = (yargs: Argv): Argv =>
  yargs
    .commandDir('content-item', YargsCommandBuilderOptions)
    .options(configureCommandOptions)
    .config('config', readConfig)
    .demandCommand()
    .help();

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const handler = (): void => {};
