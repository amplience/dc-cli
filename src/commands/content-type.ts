import { Argv } from 'yargs';
import { readConfig } from '../cli';
import YargsCommandBuilderOptions from '../common/yargs/yargs-command-builder-options';
import { configureCommandOptions } from './configure';

export const command = 'content-type';

export const desc = 'Content Type';

export const builder = (yargs: Argv): Argv =>
  yargs
    .commandDir('content-type', YargsCommandBuilderOptions)
    .options(configureCommandOptions)
    .config('config', readConfig)
    .demandCommand()
    .help();

export const handler = (): void => {
  /* do nothing */
};
