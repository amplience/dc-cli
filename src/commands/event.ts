import { Argv } from 'yargs';
import { readConfig } from '../cli';
import YargsCommandBuilderOptions from '../common/yargs/yargs-command-builder-options';
import { configureCommandOptions } from './configure';

export const command = 'event';

export const desc = 'Event';

export const builder = (yargs: Argv): Argv =>
  yargs
    .commandDir('event', YargsCommandBuilderOptions)
    .options(configureCommandOptions)
    .config('config', readConfig)
    .demandCommand()
    .help();
