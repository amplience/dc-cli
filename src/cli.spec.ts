import * as cli from './cli';
import Yargs from 'yargs/yargs';
import { configureCommandOptions } from './commands/configure';
import YargsCommandBuilderOptions from './common/yargs/yargs-command-builder-options';

jest.mock('./commands/configure');

describe('cli', (): void => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  const captureConsole = async (methodToWrap: () => void): Promise<string> => {
    let buffer = '';
    jest.spyOn(console, 'error').mockImplementation((output: string) => (buffer = `${buffer}${output}`));
    await methodToWrap();
    // replace config file and script entry point
    return buffer;
  };

  it('should configure yarg instance', async (): Promise<void> => {
    const argv = Yargs(process.argv.slice(2));
    const spyOptions = jest.spyOn(argv, 'options').mockReturnThis();
    const spyConfig = jest.spyOn(argv, 'config').mockReturnThis();
    const spyCommandDir = jest.spyOn(argv, 'commandDir').mockReturnThis();
    const spyDemandCommand = jest.spyOn(argv, 'demandCommand').mockReturnValue(argv);

    await cli.default(argv);

    expect(spyOptions).toHaveBeenCalledWith(configureCommandOptions);
    expect(spyConfig).toHaveBeenCalledWith('config', expect.any(Function));
    expect(spyCommandDir).toHaveBeenCalledWith('./commands', YargsCommandBuilderOptions);
    expect(spyDemandCommand).toHaveBeenCalledWith(1, 'Please specify at least one command');
  });

  it('should create a yarg instance if one is not supplied', async () => {
    configureCommandOptions.config.default = 'config.json';
    const processExitSpy = jest.spyOn(process, 'exit').mockImplementation();

    const buffer = await captureConsole(async () => {
      await cli.default();
    });

    expect(buffer).toMatchSnapshot();
    expect(processExitSpy).not.toHaveBeenCalled();
  });
});
