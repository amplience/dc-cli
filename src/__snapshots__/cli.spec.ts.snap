// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should create a yarg instance if one is not supplied 1`] = `
"dc-cli <command>

Commands:
  dc-cli configure            Saves the configuration options to a file
  dc-cli content-repository   Content Repository
  dc-cli content-type-schema  Content Type Schema
  dc-cli content-type         Content Type

Options:
  --help          Show help                                            [boolean]
  --version       Show version number                                  [boolean]
  --clientId                                                 [string] [required]
  --clientSecret                                             [string] [required]
  --hubId                                                    [string] [required]
  --config        Path to JSON config file     [string] [default: \\"config.json\\"]
Error: Please specify at least one command"
`;
